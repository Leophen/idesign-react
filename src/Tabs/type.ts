export interface TabsProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 选项卡风格类型
   * @default normal
   */
  theme?: 'normal' | 'card';
  /**
   * 固定选中值（受控）
   */
  active?: string | number;
  /**
   * 默认选中值（非受控）
   */
  defaultActive?: string | number;
  /**
   * 全局禁用
   * @default false
   */
  disabled?: boolean;
  /**
   * 点击选项卡时触发
   */
  onChange?: (value: string | number) => void;
}

export interface TabsItemProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 单项值
   */
  value: string | number;
  /**
   * 点击选项卡时触发
   */
  onClick?: (value?: string | number) => void;
}

export interface TabsItemAddProps extends TabsItemProps {
  /**
   * 选项卡风格类型
   * @default normal
   */
  theme?: 'normal' | 'card';
  /**
   * 当前选中值
   */
  active?: string | number;
  /**
   * 索引值
   */
  index?: number;
  /**
   * 全局禁用
   * @default false
   */
  disabled?: boolean;
  /**
   * 点击选项卡时触发
   */
  onChange?: (value: string | number) => void;
}

export interface TabsContextValue {
  inject: (props: TabsItemProps) => TabsItemAddProps;
}
