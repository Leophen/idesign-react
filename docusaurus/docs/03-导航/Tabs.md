---
title: Tabs 标签页
---

明示任务流程和当前完成程度，引导用户按照步骤完成任务。

## 引入

```js
import {Tabs} from 'idesign';

const {TabPane} = Tabs;
```

## 基本用法

```jsx live fffx
<Tabs defaultActiveKey='1'>
    <TabPane key='1' tab='Tab 1'>
        内容1
    </TabPane>
    <TabPane key='2' tab='Tab 2' disabled>
        内容2
    </TabPane>
    <TabPane key='3' tab='Tab 3'>
        内容3
    </TabPane>
</Tabs>
```

## 带图标

```jsx live fffx
function App() {
    function callback(key) {
        console.log(key);
    }

    return (
        <Tabs defaultActiveKey="1" onChange={callback}>
            <TabPane
                tab={(
                    <>
                        <OutlinedVidicon />Tab1
                    </>
                )}
                key="1"
            >
                内容1
            </TabPane>
            <TabPane
                tab={(
                    <>
                        <OutlinedVidicon />Tab2
                    </>
                )}
                key="2"
            >
                内容2
            </TabPane>
        </Tabs>
    );
}
```

## 卡片式

```jsx live fffx
<Tabs defaultActiveKey='1' type='card'>
    <TabPane key='1' tab='Tab 1'>
        内容1
    </TabPane>
    <TabPane key='2' tab='Tab 2'>
        内容2
    </TabPane>
    <TabPane key='3' tab='Tab 3'>
        内容3
    </TabPane>
</Tabs>
```

## 动态增减页签

```jsx live fff
function App() {
    const initialPanes = [
        {tab: 'Tab', content: '内容1', key: '1'},
        {tab: 'Tab', content: '内容2', key: '2'},
        {tab: 'Tab', content: '内容3', key: '3'}
    ];

    const [activeKey, setActiveKey] = useState(initialPanes[0].key);
    const [panes, setPanes] = useState(initialPanes);
    const newTabIndex = useRef(0);

    const onChange = newActiveKey => {
        setActiveKey(newActiveKey);
    };

    const add = () => {
        const newActiveKey = `newTab${newTabIndex.current++}`;
        const newPanes = [...panes];
        newPanes.push({tab: 'New Tab', content: 'Content of new Tab', key: activeKey });
        setPanes(newPanes);
        setActiveKey(newActiveKey);
    };

    const remove = targetKey => {
        let newActiveKey = activeKey;
        let lastIndex = -1;
        panes.forEach((item, i) => {
            if (item.key === targetKey) {
                lastIndex = i - 1;
            }
        });
        const newPanes = panes.filter(item => item.key !== targetKey);
        if (newPanes.length && newActiveKey === targetKey) {
            if (lastIndex >= 0) {
                newActiveKey = newPanes[lastIndex].key;
            }
            else {
                newActiveKey = newPanes[0].key;
            }
        }
        setPanes(newPanes);
        setActiveKey(newActiveKey);
    };

    const onEdit = (targetKey, action) => {
        if (action === 'add') {
            add();
        }
        else {
            remove(targetKey);
        }
    };

    return (
        <Tabs
            type="editable-card"
            onChange={onChange}
            activeKey={activeKey}
            onEdit={onEdit}
        >
            {panes.map(pane => (
                <TabPane tab={pane.tab} key={pane.key} closable={pane.closable}>
                    {pane.content}
                </TabPane>
            ))}
        </Tabs>
    );
}
```

## API

| 参数               | 说明                                                              | 类型                                                                                     | 默认值                                                         |
| ------------------ | ----------------------------------------------------------------- | ---------------------------------------------------------------------------------------- | -------------------------------------------------------------- |
| activeKey          | 当前激活 tab 面板的 key                                           | `string`                                                                                 | `--`                                                           |
| addIcon            | 自定义添加按钮                                                    | `ReactNode`                                                                              | `+`                                                            |
| animated           | 是否使用动画切换 Tabs                                             | `boolean`                                                                                | `{ inkBar: boolean, tabPane: boolean }`                        |
| centered           | 标签居中展示                                                      | `boolean`                                                                                | `false`                                                        |
| defaultActiveKey   | 初始化选中面板的 key，如果没有设置 activeKey                      | `string`                                                                                 | `第一个面板`                                                   |
| hideAdd            | 是否隐藏加号图标，在 type="editable-card" 时有效                  | `boolean`                                                                                | `false`                                                        |
| renderTabBar       | 替换 TabBar，用于二次封装标签头                                   | `(props: DefaultTabBarProps, DefaultTabBar: React.ComponentClass) => React.ReactElement` | `--`                                                           |
| tabBarExtraContent | tab bar 上额外的元素                                              | `{left?: ReactNode, right?: ReactNode}`                                                  | `true`                                                         |
| tabBarGutter       | tabs 之间的间隙                                                   | `number`                                                                                 | `--`                                                           |
| tabBarStyle        | tab bar 的样式对象                                                | `object`                                                                                 | `--`                                                           |
| type               | 页签的基本样式，可选 `line`、`card` `editable-card` 类型          | `string`                                                                                 | `line`                                                         |
| onChange           | 切换面板的回调                                                    | `function(activeKey) {}`                                                                 | `--`                                                           |
| onEdit             | 新增和删除页签的回调，在 type="editable-card" 时有效              | `(targetKey, action): void`                                                              | `--`                                                           |
| onTabClick         | tab 被点击的回调                                                  | `function(key: string, event: MouseEvent)`                                               | `--`                                                           |
| onEditTabNav       | 双击tabNav后被编辑的回调（只允许 tab 为 string 的 tabNav 被编辑） | `function(panes)`                                                                        | `[{ tab: string, content: string \| ReactNode, key: string }]` |

## Tabs.TabPane

| 参数        | 说明                                               | 类型                | 默认值  |
| ----------- | -------------------------------------------------- | ------------------- | ------- |
| closeIcon   | 自定义关闭图标，在 type="editable-card"时有效      | `ReactNode`         | `--`    |
| forceRender | 被隐藏时是否渲染 DOM 结构                          | `boolean`           | `false` |
| key         | 对应 activeKey                                     | `string`            | `--`    |
| tab         | 选项卡头显示文字                                   | `string〡ReactNode` | `--`    |
| max         | 选项卡头显示最多显示文字字数，tab 为 string 时有效 | `number`            | `--`    |
| disabled    | 选项卡头不可点                                     | `boolean`           | `false` |
