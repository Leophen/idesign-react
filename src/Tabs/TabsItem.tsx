import React, { useContext } from 'react';
import classNames from 'classnames';
import './index.scss';
import { TabsContext } from './index'
import { TabsItemAddProps, TabsItemProps } from './type';

const TabsItem: React.FC<TabsItemProps> = (props) => {
  // 从 Tabs Context 注入全局属性
  const context = useContext(TabsContext);
  const newProps: TabsItemAddProps = context ? context.inject(props) : props;

  const {
    children,
    className,
    style,
    value,
    onClick,
    // 以下为 context 传入或结合
    theme,
    index,
    active,
    disabled = false,
    onChange,
    ...restProps
  } = newProps;

  const itemValue = value || index

  const handleClickTab = () => {
    !disabled && onClick?.(itemValue)
  }

  return (
    <div
      className={classNames(
        'i-tabs-item',
        itemValue === active && 'i-tabs-item__active',
        theme === 'card' && 'i-tabs-item__card',
        disabled && 'i-tabs-item__disabled',
        className
      )}
      data-active={itemValue === active}
      data-disabled={disabled}
      style={{ ...style }}
      onClick={handleClickTab}
      {...restProps}
    >
      {children}
    </div>
  );
};

export default TabsItem;
