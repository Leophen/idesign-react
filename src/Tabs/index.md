---
nav:
  title: 组件
  path: /components
group:
  title: 导航组件
  order: 2
order: 2
---

# Tabs 选项卡

用于在同个页面下进行快速切换的选项卡。

## 基本用法

使用 `Tabs` 包裹菜单的每一项 `Tabs.Item`：

```tsx
import React, { useState } from 'react'
import { Tabs } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState(1)

  const handleChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <>
      <h4>无默认值</h4>
      <Tabs>
        <Tabs.Item>选项卡1</Tabs.Item>
        <Tabs.Item>选项卡22</Tabs.Item>
        <Tabs.Item>选项卡333</Tabs.Item>
      </Tabs>
      <h4>有默认值</h4>
      <Tabs defaultActive={value}>
        <Tabs.Item>选项卡1</Tabs.Item>
        <Tabs.Item>选项卡22</Tabs.Item>
        <Tabs.Item>选项卡333</Tabs.Item>
      </Tabs>
      <h4>有受控值</h4>
      <Tabs active={value}>
        <Tabs.Item>选项卡1</Tabs.Item>
        <Tabs.Item>选项卡22</Tabs.Item>
        <Tabs.Item>选项卡333</Tabs.Item>
      </Tabs>
      <h4>一般用法</h4>
      <Tabs active={value} onChange={handleChange}>
        <Tabs.Item>选项卡1</Tabs.Item>
        <Tabs.Item>选项卡22</Tabs.Item>
        <Tabs.Item>选项卡333</Tabs.Item>
      </Tabs>
    </>
  )
}

export default App
```

## 卡片风格

通过 `theme` 属性指定卡片风格：

```tsx
import React, { useState } from 'react'
import { Tabs } from 'idesign-react'

const App = () => {
  return (
    <Tabs defaultActive="2" theme="card">
      <Tabs.Item value="1">选项卡1</Tabs.Item>
      <Tabs.Item value="2">选项卡22</Tabs.Item>
      <Tabs.Item value="3">选项卡333</Tabs.Item>
    </Tabs>
  )
}

export default App
```

## 禁用状态

通过 `disabled` 属性指定选项卡禁用：

```tsx
import React, { useState } from 'react'
import { Tabs } from 'idesign-react'

const App = () => {
  return (
    <>
      <h4>全局禁用</h4>
      <Tabs disabled>
        <Tabs.Item>选项卡1</Tabs.Item>
        <Tabs.Item>选项卡22</Tabs.Item>
        <Tabs.Item>选项卡333</Tabs.Item>
      </Tabs>
      <h4>单项禁用</h4>
      <Tabs>
        <Tabs.Item>选项卡1</Tabs.Item>
        <Tabs.Item disabled>选项卡22</Tabs.Item>
        <Tabs.Item>选项卡333</Tabs.Item>
      </Tabs>
      <h4>卡片风格全局禁用</h4>
      <Tabs theme="card" disabled>
        <Tabs.Item>选项卡1</Tabs.Item>
        <Tabs.Item>选项卡22</Tabs.Item>
        <Tabs.Item>选项卡333</Tabs.Item>
      </Tabs>
      <h4>卡片风格单项禁用</h4>
      <Tabs theme="card">
        <Tabs.Item>选项卡1</Tabs.Item>
        <Tabs.Item disabled>选项卡22</Tabs.Item>
        <Tabs.Item>选项卡333</Tabs.Item>
      </Tabs>
    </>
  )
}

export default App
```

## Tabs API

<API hideTitle />

## TabsItem API

| 属性      | 说明             | 类型                               | 默认值  |
| --------- | ---------------- | ---------------------------------- | ------- |
| children  | 按钮内容         | `ReactNode`                        | `--`      |
| className | 类名             | `string`                           | `--`      |
| style     | 自定义样式       | `CSSProperties`                    | `--`      |
| value     | 单项值           | `string〡number`                   | `--`      |
| disabled  | 单项禁用         | `boolean`                          | `false` |
| onClick   | 点击单项卡时触发 | `(value?: string〡number) => void` | `--`      |
