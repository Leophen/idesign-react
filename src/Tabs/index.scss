.i-tabs {
  position: relative;
  display: flex;
  box-sizing: border-box;
  font-size: 14px;
  border-bottom: 1px solid var(--i-border);
}

.i-tabs-item {
  position: relative;
  display: flex;
  align-items: center;
  box-sizing: border-box;
  height: 40px;
  padding: 0 12px;
  color: var(--i-font-2);
  cursor: pointer;
  &.i-tabs-item__active {
    color: var(--i-primary-txt);
  }
  &.i-tabs-item__disabled {
    color: var(--i-font-disabled);
    cursor: not-allowed;
  }
  &.i-tabs-item__card {
    border-top: 1px solid var(--i-border);
    border-left: 1px solid var(--i-border);
    &:first-of-type {
      border-radius: 4px 0 0 0;
    }
    &:last-of-type {
      border-right: 1px solid var(--i-border);
      border-radius: 0 4px 0 0;
    }
    &.i-tabs-item__active {
      background: var(--i-bg);
      &::after {
        content: '';
      }
    }
    &::after {
      position: absolute;
      bottom: -1px;
      left: 0;
      display: block;
      width: 100%;
      height: 1px;
      background: var(--i-bg);
    }
  }
}

.i-tabs__bar {
  position: absolute;
  bottom: 0px;
  height: 2px;
  background: var(--i-primary-txt);
  border-radius: 1px;
  transition: all 0.2s ease-out;
}
