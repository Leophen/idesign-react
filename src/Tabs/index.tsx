import React, { useEffect, useRef, useState } from 'react';
import classNames from 'classnames';
import './index.scss';
import useDefault from '../hooks/useDefault';
import TabsItem from './TabsItem';
import { TabsContextValue, TabsItemProps, TabsProps } from './type';

export const TabsContext = React.createContext<TabsContextValue>(null as any);

const Tabs: React.FC<TabsProps> & { Item: React.ElementType } = (props) => {
  let defaultVal
  React.Children.map(props.children, (child, index) => {
    index === 0 && (defaultVal = (child as any).props.value || 0)
  })

  const {
    children = '',
    className,
    style,
    active,
    defaultActive = defaultVal,
    disabled = false,
    theme = 'normal',
    onChange,
    ...restProps
  } = props;

  const [innerActive, setInnerActive] = useDefault(active, defaultActive, onChange);

  const tabsRef = useRef<HTMLDivElement>(null)
  const [tabsRefLeft, setTabsRefLeft] = useState(0)
  useEffect(() => {
    const curTabsRefLeft = tabsRef.current?.getBoundingClientRect().left || 0
    setTabsRefLeft(curTabsRefLeft)
    updateBarPosition(curTabsRefLeft)
  }, [])

  const [bar, setBar] = useState({
    width: 0,
    left: 0
  })

  const updateBarPosition = (parentLeft: number) => {
    tabsRef.current?.childNodes.forEach(item => {
      if (
        (item as HTMLElement).dataset.active === 'true' &&
        (item as HTMLElement).dataset.disabled !== 'true'
      ) {
        const tabRect = (item as HTMLElement).getBoundingClientRect()
        bar.left = tabRect.left - parentLeft
        bar.width = tabRect.width
        setBar({ ...bar })
      }
    })
  }

  useEffect(() => {
    tabsRefLeft !== 0 && updateBarPosition(tabsRefLeft)
  }, [innerActive])

  // 注入每一项的 context
  const context: TabsContextValue = {
    // 将 props 注入每一项子节点的方法
    inject: (singleTabsProps: TabsItemProps) => {
      return {
        theme,
        active: innerActive,
        disabled,
        onClick(value?: string | number) {
          setInnerActive(value)
        },
        ...singleTabsProps,
      };
    },
  };

  return (
    <TabsContext.Provider value={context}>
      <div
        className={classNames(
          'i-tabs',
          className
        )}
        ref={tabsRef}
        style={{ ...style }}
        {...restProps}
      >
        {React.Children.map(children, (child, index) => {
          if (!React.isValidElement(child)) {
            return null;
          }
          const childProps = {
            index,
            ...child.props
          };
          return React.cloneElement(child, childProps);
        })}
        {theme === 'normal' && (
          <div
            className="i-tabs__bar"
            style={{
              width: bar.width,
              left: bar.left
            }}
          />
        )}
      </div>
    </TabsContext.Provider>
  );
};

Tabs.Item = TabsItem;

Tabs.displayName = 'Tabs';
TabsItem.displayName = 'TabsItem';

export default Tabs;
