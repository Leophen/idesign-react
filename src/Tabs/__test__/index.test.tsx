import { render, fireEvent } from '@testing-library/react';
import Tabs from '../index';

describe('Tabs 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(
      <Tabs>
        <Tabs.Item>选项卡1</Tabs.Item>
        <Tabs.Item>选项卡2</Tabs.Item>
        <Tabs.Item>选项卡3</Tabs.Item>
      </Tabs>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('active', () => {
    const { getByTestId } = render(
      <Tabs active={1}>
        <Tabs.Item>选项卡1</Tabs.Item>
        <Tabs.Item data-testid='test-item'>选项卡2</Tabs.Item>
        <Tabs.Item>选项卡3</Tabs.Item>
      </Tabs>
    );
    expect(getByTestId('test-item')).toHaveClass('i-tabs-item__active');
  });

  it('defaultActive', () => {
    const { getByTestId } = render(
      <Tabs defaultActive={1}>
        <Tabs.Item>选项卡1</Tabs.Item>
        <Tabs.Item data-testid='test-item'>选项卡2</Tabs.Item>
        <Tabs.Item>选项卡3</Tabs.Item>
      </Tabs>
    );
    expect(getByTestId('test-item')).toHaveClass('i-tabs-item__active');
  });

  it('onChange', () => {
    const changeFn = jest.fn();
    const { getByTestId } = render(
      <Tabs onChange={changeFn}>
        <Tabs.Item>选项卡1</Tabs.Item>
        <Tabs.Item data-testid='test-item'>选项卡2</Tabs.Item>
        <Tabs.Item>选项卡3</Tabs.Item>
      </Tabs>
    );
    fireEvent.click(getByTestId('test-item'));
    expect(changeFn).toHaveBeenCalled();
  });

  it('theme', () => {
    const { container } = render(
      <Tabs theme="card">
        <Tabs.Item>选项卡1</Tabs.Item>
        <Tabs.Item>选项卡2</Tabs.Item>
        <Tabs.Item>选项卡3</Tabs.Item>
      </Tabs>
    );
    expect(container.firstChild.firstChild).toHaveClass('i-tabs-item__card');
  });

  it('disabled', () => {
    const changeFn = jest.fn();
    const { getByTestId } = render(
      <Tabs disabled>
        <Tabs.Item>选项卡1</Tabs.Item>
        <Tabs.Item data-testid='test-item'>选项卡2</Tabs.Item>
        <Tabs.Item>选项卡3</Tabs.Item>
      </Tabs>
    );
    fireEvent.click(getByTestId('test-item'));
    expect(changeFn).not.toHaveBeenCalled();
  });
});

