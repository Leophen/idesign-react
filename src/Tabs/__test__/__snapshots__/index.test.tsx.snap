// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-tabs"
  >
    <div
      class="i-tabs-item i-tabs-item__active"
      data-active="true"
      data-disabled="false"
    >
      选项卡1
    </div>
    <div
      class="i-tabs-item"
      data-active="false"
      data-disabled="false"
    >
      选项卡2
    </div>
    <div
      class="i-tabs-item"
      data-active="false"
      data-disabled="false"
    >
      选项卡3
    </div>
    <div
      class="i-tabs__bar"
      style="width: 0px; left: 0px;"
    />
  </div>
</DocumentFragment>
`;
